stellargraph/core/element_data.py

```diff
@@ -121,40 +121,60 @@ class ElementData:
 
     Args:
         ids (sequence): the IDs of each element
-        type_starts (list of tuple of type name, int): the starting iloc of the elements of each type within ``shared``
+        type_info (list of tuple of type name, numpy array): the associated feature vectors of each type, where the size of the first dimension defines the elements of that type
     """
 
-    def __init__(self, ids, type_starts):
-        if not isinstance(type_starts, list):
+    def __init__(self, ids, type_info):
+        if not isinstance(type_info, list):
             raise TypeError(
-                f"type_starts: expected list, found {type(type_starts).__name__}"
+                f"type_info: expected list, found {type(type_info).__name__}"
             )
 
         type_ranges = {}
-        type_stops = type_starts[1:] + [(None, len(ids))]
-        consecutive_types = zip(type_starts, type_stops)
-        for idx, ((type_name, start), (_, stop)) in enumerate(consecutive_types):
-            if idx == 0 and start != 0:
-                raise ValueError(
-                    f"type_starts: expected first type ({type_name!r}) to start at index 0, found start {start}"
-                )
-            if start > stop:
+        features = {}
+        all_types = []
+        type_sizes = []
+
+        rows_so_far = 0
+
+        # validation
+        for type_name, data in type_info:
+            if not isinstance(data, np.ndarray):
                 raise TypeError(
-                    f"type_starts (for {type_name!r}): expected valid type range, found start ({start}) after stop ({stop})"
+                    f"type_info (for {type_name!r}): expected numpy array, found {type(data).__name__}"
                 )
+
+            if len(data.shape) != 2:
+                raise ValueError(
+                    f"type_info (for {type_name!r}): expected 2 dimensions, found {len(data.shape)}"
+                )
+
+            rows, _columns = data.shape
```

Comment:
I think `_columns` is unused here
```suggestion
            rows = data.shape[0]
```

Reply:
It was on purpose in the original code (that's what the leading _ conveys, in my mind), trying to express the rank-2 nature of `data`... however, that was all changed by the multidimensional feature change in #1561 that's now merged, and I guess it'll be resolved by the merge with develop.